core/client/src/client.rs

```diff
@@ -1477,21 +1355,202 @@ where
 	}
 }
 
-impl<B, E, Block, RA> ChainHead<Block> for Client<B, E, Block, RA>
+/// Implement Longest Chain Select implementation
+/// where 'longest' is defined as the highest number of blocks
+pub struct LongestChain<B, Block> {
+	backend: Arc<B>,
+	import_lock: Arc<Mutex<()>>,
+	_phantom: PhantomData<Block>
+}
+
+impl<B, Block> Clone for LongestChain<B, Block> {
+	fn clone(&self) -> Self {
+		let backend = self.backend.clone();
+		let import_lock = self.import_lock.clone();
+		LongestChain {
+			backend,
+			import_lock,
+			_phantom: Default::default()
+		}
+	}
+}
+
+impl<B, Block> LongestChain<B, Block>
 where
 	B: backend::Backend<Block, Blake2Hasher>,
-	E: CallExecutor<Block, Blake2Hasher>,
 	Block: BlockT<Hash=H256>,
 {
+	/// Instantiate a new LongestChain for Backend B
+	pub fn new(backend: Arc<B>, import_lock: Arc<Mutex<()>>) -> Self {
+		LongestChain {
+			backend,
+			import_lock,
+			_phantom: Default::default()
+		}
+	}
+
 	fn best_block_header(&self) -> error::Result<<Block as BlockT>::Header> {
-		Client::best_block_header(self)
+		let info : ChainInfo<Block> = match self.backend.blockchain().info() {
+			Ok(i) => i,
+			Err(e) => return Err(error::Error::from_blockchain(Box::new(e)))
+		};
+		Ok(self.backend.blockchain().header(BlockId::Hash(info.best_hash))?
+			.expect("Best block header must always exist"))
```

Comment:
It would be good to have a test case that checks that this is true, especially on genesis.

Reply:
I was looking into this, but I don't think I want to introduce this in this PR.

As genesis is chain-specific, this test could be done against our test-client (which is a non-test) or as an integration test on `node` or other implementations. However, for the later, we don't really have any of these tests now, nor have we thought thoroughly enough how we want to support for, implement and run these tests (for e.g. I'd love to have a bunch of test in the template, that ensure assumptions like this for the genesis and service config developer might write and give them hints as to how to fix things if the test complain).

Adding this specific tests would require a bigger degree of changes in `node`/`node-template` to facilitate for this, but I don't think we have thought well enough of what we want there to add this hastily at this point.